moshi-adapters/src/main/java/com/squareup/moshi/adapters/PolymorphicJsonAdapterFactory.kt

```diff
@@ -0,0 +1,274 @@
+ *
+ * If an unknown type is encountered when encoding:
+ *  * If [withFallbackJsonAdapter] is used, then the `fallbackJsonAdapter.toJson(writer, value)` result will be returned.
+ *  * Otherwise a [IllegalArgumentException] will be thrown.
+ *
+ * If the same subtype has multiple labels the first one is used when encoding.
+ */
+public class PolymorphicJsonAdapterFactory<T> internal constructor(
+  private val baseType: Class<T>,
+  private val labelKey: String,
+  private val labels: List<String>,
+  private val subtypes: List<Type>,
+  private val fallbackJsonAdapter: JsonAdapter<Any>?
+) : Factory {
+  /** Returns a new factory that decodes instances of `subtype`. */
+  public fun withSubtype(subtype: Class<out T>, label: String): PolymorphicJsonAdapterFactory<T> {
+    require(!labels.contains(label)) { "Labels must be unique." }
+    val newLabels = buildList {
+      addAll(labels)
+      add(label)
+    }
+    val newSubtypes = buildList {
+      addAll(subtypes)
+      add(subtype)
+    }
+    return PolymorphicJsonAdapterFactory(
+      baseType = baseType,
+      labelKey = labelKey,
+      labels = newLabels,
+      subtypes = newSubtypes,
+      fallbackJsonAdapter = fallbackJsonAdapter
+    )
+  }
+
+  /**
+   * Returns a new factory that with default to `fallbackJsonAdapter.fromJson(reader)` upon
+   * decoding of unrecognized labels.
+   *
+   * The [JsonReader] instance will not be automatically consumed, so make sure to consume
+   * it within your implementation of [JsonAdapter.fromJson]
+   */
+  public fun withFallbackJsonAdapter(
+    fallbackJsonAdapter: JsonAdapter<Any>?
+  ): PolymorphicJsonAdapterFactory<T> {
+    return PolymorphicJsonAdapterFactory(
+      baseType = baseType,
+      labelKey = labelKey,
+      labels = labels,
+      subtypes = subtypes,
+      fallbackJsonAdapter = fallbackJsonAdapter
+    )
+  }
+
+  /**
+   * Returns a new factory that will default to `defaultValue` upon decoding of unrecognized
+   * labels. The default value should be immutable.
+   */
+  public fun withDefaultValue(defaultValue: T?): PolymorphicJsonAdapterFactory<T> {
+    return withFallbackJsonAdapter(buildFallbackJsonAdapter(defaultValue))
+  }
+
+  private fun buildFallbackJsonAdapter(defaultValue: T?): JsonAdapter<Any> {
+    return object : JsonAdapter<Any>() {
+      override fun fromJson(reader: JsonReader): Any? {
+        reader.skipValue()
+        return defaultValue
+      }
+
+      override fun toJson(writer: JsonWriter, value: Any?) {
+        throw IllegalArgumentException(
+          "Expected one of $subtypes but found $value, a ${value?.javaClass}. Register this subtype."
+        )
+      }
+    }
+  }
+
+  override fun create(type: Type, annotations: Set<Annotation?>, moshi: Moshi): JsonAdapter<*>? {
+    if (type.rawType != baseType || annotations.isNotEmpty()) {
+      return null
+    }
+    val jsonAdapters: List<JsonAdapter<Any>> = List(subtypes.size) { i ->
+      moshi.adapter(subtypes[i])
+    }
```

Comment:
nit: is this `List` construction more idiomatic than `subtypes.map { moshi.adapter(it) }`?